### ISSUE TITLE
cannot get and set a signal within the same reactive context

### ISSUE THREAD

**Issue by jquesada2016** [gbj/leptos#83]
The following example panics at `already borrowed`.

```rust
use leptos::*;
use wasm_bindgen::prelude::*;

#[wasm_bindgen(start)]
pub fn start() {
  console_error_panic_hook::set_once();

  mount_to_body(view_fn);
}

fn view_fn(cx: Scope) -> impl Mountable {
  let (a, set_a) = create_signal(cx, false);

  create_effect(cx, move |_| {
    if !a() {
      set_a(true);
    }
  });

  view! { cx,
    <h1>"Hello"</h1>
  }
}
```

**Comment by gbj** [comment]
Yeah so this is working as designed, but perhaps should be better documented. This is why I give the advice to avoid setting signals within effects, but rather to derive a signal—it's very easy to create an infinite loop if you do this, and it's basically useless. Your example defines `a` as `false` but then immediately, synchronously sets it to `true`.

Is there a less simple use case you can provide where you might actually want to do this?

Logically it's the equivalent of

```rust
let a = RefCell::new(false);

if !a.borrow() {
  *a.borrow_mut() = true;
}
```

and panics for more or less the same reason.

**Comment by jquesada2016** [comment]
Yes, of course, this was an min-repro.

I ran into this in the following situation.

I have a list of users, that can potentially change at any time. Let's call this signal `users: Vec<User>`. I need to create a bool for each user to see if they are selected or not. Let's call this signal `selected_users: Vec<bool>`. Since the number of users can change, I need to be able to update the size of `selected_users`, but cannot do this without panicking.

I know one solution would be to move the logic which adjusts the size of the `selected_users` into it's own effect, but I need this effect to be guaranteed to run before another effect, otherwise I'll break the size invariant, and there's no current way to make sure an effect runs before another.

Here's the above example in more concrete code:

```rust
let users_view = move || {
  let users: Vec<User> = users(); // Signal
  
  if users.len() != selected_users().len() {
    set_selected_users(vec![false; users.len()]);
  }
  
  users
    .into_iter()
    .map(|u| view! { cx, /* view stuff */})
    .collect::<Vec<_>>()
};
```

**Comment by gbj** [comment]
The `update` function (instead of the `set` function) gives you a mutable reference to the current value: since `users()` is the only dependency the below should work fine, shouldn't it?

```rust
let users_view = move || {
  let users: Vec<User> = users(); // Signal
  
  set_selected_users.update(|selected_users| {
    if users.len() != selected_users.len() {
      selected_users. /* whatever the logic is to trim it */
    }
  });
  
  users
    .into_iter()
    .map(|u| view! { cx, /* view stuff */})
    .collect::<Vec<_>>()
};
```

**Comment by gbj** [comment]
I'd probably recommend either bundling selected users and users into one `Vec<(User, bool)>` but I think the solution I posted should work.

**Comment by jquesada2016** [comment]
Yes, the `update` method you suggested above would totally work, but if I may ask, how does it actually work in the implementation?

The example you provided of

```rust
let a = RefCell::new(false);

if !*a.borrow() {
  *a.borrow_mut() = true;
}
```

would not actually panic in practice, because the immutable borrow is dropped as soon as the `if` is evaluated. I would expect the same behavior to persist within an effect; that is, when you get a signal, it borrows it mutably and immediately releases the borrow, therefore, when a subsequent mutable borrow occurs, it's the sole reference, and therefore will not panic.

**Comment by gbj** [comment]
Oh you're right of course, about the RefCell parallel.

The reason it doesn't work in the example you gave is

```rust
create_effect(cx, move |_| {
  if !a() { // now the effect depends on 
    set_a(true);
  }
});
```

**Comment by gbj** [comment]
It panics because running an effect, at the moment, mutably borrows the effect function, because `create_effect()` takes `FnMut(Option<&T>) -> T`.  `set_a()` triggers the effect to run again, but the effect function is already borrowed mutably. 

So the rule is more like "you can't trigger an effect to re-run from within an effect."

***

It's possible that effect functions don't need to be `FnMut`; they can take `Fn` and you can always wrap a value being moved into it in a `RefCell` if you need to mutate it. 

Let me try to make this change and see if it a) fixes the bug and b) still lets everything else work.

And thanks for pushing on this!

**Comment by gbj** [comment]
This should be resolved now by #90. Thanks again.
